csrc/executor.cpp

```diff
@@ -645,14 +645,331 @@ std::pair<std::vector<int64_t>, std::vector<int64_t>> inferShapeOfOutput(
   return inferShape(tv, symbolic_sizes, expand_flags, expr_eval);
 }
 
-// Allocate output tensos for a given kernel. Outputs may alias inputs, in
+namespace {
+
+class ForwardTraverseFromAllocToRFactor {
+  at::Tensor tensor_;
+  TensorView* tv_;
+  ExpressionEvaluator& ee_;
+  std::list<IterDomain*>& frontier_;
+
+  // Forward traverse split from allocation to rFactor. Needs to, for example,
+  // view tensor with shape [..., 15, ...] as [..., 3, 5, ...]
+  void handle(Split* split) {
+    auto in = split->in();
+    auto inner = split->inner();
+    auto outer = split->outer();
+    auto factor = ee_.evaluate(split->factor())->as<int64_t>();
+    auto in_it = std::find(frontier_.begin(), frontier_.end(), in);
+    // TORCH_INTERNAL_ASSERT(in_it != frontier_.end());
+    if (in_it == frontier_.end()) {
+      // TODO: We should get rid of this return and enable the above assert.
+      // Note [Allocation domain on both side of rFactor]
+      // For cases where the allocation domain is on both side of rFactor, for
+      // example, in Tensor3d_To_NHWC4d_FwdBwd_CUDA:
+      // [alloc,root]   [alloc,root]           [root]
+      //          \     /                      /    |
+      //         [rFactor]                  split   [rFactor]
+      //                                    /  \         |
+      //                      [alloc,rFactor] [rFactor]  |
+      //                                             \   |
+      //                                             [alloc]
+      // I have no idea why StmtSort::getExprsBetween is not returning the
```

Comment:
What exprs are returned?

Reply:
The split in the above figure is returned, probably because one output of the split is both in alloc and in rfactor?

Reply:
I think that's because there's nothing to stop from the inner split output to get traversed to the root. I believe there's a note in some of the dependency API functions.